File: 0767.ReorganizeString.py
class Solution(object):
    def reorganizeString(self, S):
        """
        :type S: str
        :rtype: str
        """
        heap = [(-S.count(ch), ch) for ch in set(S)]
        if any(-cnt > (len(S)+1)>>1 for cnt, ch in heap):
            return ''
        heapq.heapify(heap)
        ans = ''
        while len(heap) >= 2:
            cnt1, ch1 = heapq.heappop(heap)
            cnt2, ch2 = heapq.heappop(heap)
            ans = ''.join([ans, ch1, ch2])
            if cnt1+1:
                heapq.heappush(heap, (cnt1+1, ch1))
            if cnt2+1:
                heapq.heappush(heap, (cnt2+1, ch2))
        if heap:
            ans = ''.join([ans, heap[0][1]])
        return ans
        

File: 0000-0999/0767.ReorganizeString.go
func reorganizeString(S string) string {
    n := len(S)
    cnt := make([]int, 26)
    for _, ch := range S {
        cnt[ch-'a'] += 100
    }
    for i := 0; i < 26; i++ {
        cnt[i] += i
    }
    sort.Ints(cnt)
    
    ans := make([]string, n)
    i := 1
    for _, code := range cnt {
        c := code/100
        ch := string('a'+code%100)
        if c > (n+1)>>1 {
            return ""
        }
        for ; c > 0; c-- {
            if i >= n {
                i = 0
            }
            ans[i] = ch
            i += 2
        }
    }
    return strings.Join(ans, "")
}